cmd/clusterctl/client/cluster/proxy.go

```diff
@@ -102,22 +102,33 @@ func (k *proxy) ValidateKubernetesVersion() error {
 
 // GetConfig returns the config for a kubernetes client.
 func (k *proxy) GetConfig() (*rest.Config, error) {
-	config, err := k.configLoadingRules.Load()
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to load Kubeconfig")
-	}
+	var restConfig *rest.Config
 
-	configOverrides := &clientcmd.ConfigOverrides{
-		CurrentContext: k.kubeconfig.Context,
-		Timeout:        k.timeout.String(),
-	}
-	restConfig, err := clientcmd.NewDefaultClientConfig(*config, configOverrides).ClientConfig()
-	if err != nil {
-		if strings.HasPrefix(err.Error(), "invalid configuration:") {
-			return nil, errors.New(strings.Replace(err.Error(), "invalid configuration:", "invalid kubeconfig file; clusterctl requires a valid kubeconfig file to connect to the management cluster:", 1))
+	if k.kubeconfig.InCluster {
+		var err error
+		restConfig, err = rest.InClusterConfig()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create in-cluster Kubeconfig")
+		}
+	} else {
```

Comment:
Shouldn't we use the default discovery rules here, instead of providing a boolean option?

Reply:
This would come down to "it just works automatically" and discovers if it's running in-side or outside of the cluster without needing a flag?

Reply:
I didn't find a way how to use default discovery rules here

Reply:
See
https://github.com/kubernetes-sigs/cluster-api/blob/d1c9c410a364921c912e4e2163ae92754347e5f5/cmd/clusterctl/client/cluster/proxy.go#L239

and `proxy.configLoadingRules`

Reply:
I might be missing something, but default rules don't work with in-cluster config, I tried to run the current implementation in a cluster.

Reply:
What about using `GetConfigWithContext ` of controller-runtime? (that's what we're using more or less for our controllers)

The order described here seems good to me: https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/config/config.go#L63-L91